src/AggregateFunctions/AggregateFunctionDistinctJSONPaths.cpp

```diff
@@ -0,0 +1,350 @@
+    }
+
+    void serialize(WriteBuffer & buf) const
+    {
+        writeVarUInt(data.size(), buf);
+        for (const auto & path : data)
+            writeStringBinary(path, buf);
+    }
+
+    void deserialize(ReadBuffer & buf)
+    {
+        size_t size;
+        readVarUInt(size, buf);
+        if (size > DISTINCT_JSON_PATHS_MAX_ARRAY_SIZE)
+            throw Exception(ErrorCodes::TOO_LARGE_ARRAY_SIZE, "Too large array size (maximum: {}): {}", DISTINCT_JSON_PATHS_MAX_ARRAY_SIZE, size);
+
+        String path;
+        for (size_t i = 0; i != size; ++i)
+        {
+            readStringBinary(path, buf);
+            data.insert(path);
+        }
+    }
+
+    void insertResultInto(IColumn & column)
+    {
+        /// Insert paths in sorted order for better output.
+        auto & array_column = assert_cast<ColumnArray &>(column);
+        auto & string_column = assert_cast<ColumnString &>(array_column.getData());
+        std::vector<String> sorted_data(data.begin(), data.end());
+        std::sort(sorted_data.begin(), sorted_data.end());
+        for (const auto & path : sorted_data)
+            string_column.insertData(path.data(), path.size());
+        array_column.getOffsets().push_back(string_column.size());
+    }
+
+    static DataTypePtr getResultType()
+    {
+        return std::make_shared<DataTypeArray>(std::make_shared<DataTypeString>());
+    }
+};
+
+struct AggregateFunctionDistinctJSONPathsAndTypesData
+{
+    static constexpr auto name = "distinctJSONPathsAndTypes";
+
+    std::unordered_map<String, std::unordered_set<String>> data;
+
+    void add(const ColumnObject & column, size_t row_num, const std::unordered_map<String, String> & typed_paths_type_names)
+    {
+        for (const auto & [path, _] : column.getTypedPaths())
+            data[path].insert(typed_paths_type_names.at(path));
+        for (const auto & [path, dynamic_column] : column.getDynamicPathsPtrs())
+        {
+            if (!dynamic_column->isNullAt(row_num))
+                data[path].insert(dynamic_column->getTypeNameAt(row_num));
+        }
+
+        /// Iterate over paths om shared data in this row and decode the data types.
+        const auto [shared_data_paths, shared_data_values] = column.getSharedDataPathsAndValues();
+        const auto & shared_data_offsets = column.getSharedDataOffsets();
+        size_t start = shared_data_offsets[static_cast<ssize_t>(row_num) - 1];
+        size_t end = shared_data_offsets[static_cast<ssize_t>(row_num)];
+        for (size_t i = start; i != end; ++i)
+        {
+            auto path = shared_data_paths->getDataAt(i).toString();
+            auto value = shared_data_values->getDataAt(i);
+            ReadBufferFromMemory buf(value.data, value.size);
+            auto type = decodeDataType(buf);
+            /// We should not have Nulls here but let's check just in case.
+            if (!isNothing(type))
+                data[path].insert(type->getName());
```

Comment:
Or throw a logical error in case of `isNothing(type)` true, but if we do a check, then let's signal if we find something unexpected.

```suggestion
            /// We should not have Nulls here but let's check just in case.
            chassert(!isNothingType(type));
            data[path].insert(type->getName());
```